File: Cpp/0414_third_maximum_number.cpp
// https://leetcode.cn/problems/third-maximum-number/

class Solution {
public:
    int thirdMax(vector<int>& nums) {
        unordered_set<int> s(nums.begin(), nums.end());
        int max_1 = INT_MIN, max_2 = INT_MIN, max_3 = INT_MIN;
        for (auto it = s.begin(); it != s.end(); ++it) {
            if (*it > max_1) {
                max_3 = max_2;
                max_2 = max_1;
                max_1 = *it;
            }
            else if (*it > max_2) {
                max_3 = max_2;
                max_2 = *it;
            }
            else if (*it > max_3) {
                max_3 = *it;
            }
        }
        if (s.size() < 3) {
            return max_1;
        }
        else {
            return max_3;
        }
    }
};

File: Python3/0414_third_maximum_number.py
# https://leetcode.cn/problems/third-maximum-number/

class Solution:
    def thirdMax(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: int
        """

        s = set(nums)
        max_1, max_2, max_3 = float("-inf"), float("-inf"), float("-inf")
        for i in s:
            if i > max_1:
                max_3 = max_2
                max_2 = max_1
                max_1 = i
            elif i > max_2:
                max_3 = max_2
                max_2 = i
            elif i > max_3:
                max_3 = i
        if len(s) < 3:
            return max_1
        else:
            return max_3

        """
        nums_new = list(set(nums))
        n = len(nums_new)
        if n < 3:
            return max(nums_new)

        nums_new.sort()
        nums_new.pop()
        nums_new.pop()
        return nums_new[n-3]
        """